rmw_fastrtps_shared_cpp/src/rmw_wait.cpp

```diff
@@ -228,24 +206,76 @@ __rmw_wait(
     for (size_t i = 0; i < events->event_count; ++i) {
       auto event = static_cast<rmw_event_t *>(events->events[i]);
       auto custom_event_info = static_cast<CustomEventInfo *>(event->data);
-      custom_event_info->getListener()->detachCondition();
-      if (!custom_event_info->getListener()->hasEvent(event->event_type)) {
-        events->events[i] = nullptr;
+      fastdds_wait_set->detach_condition(custom_event_info->get_listener()->get_statuscondition());
+      eprosima::fastdds::dds::StatusCondition & status_condition =
+        custom_event_info->get_listener()->get_statuscondition();
+      fastdds_wait_set->detach_condition(status_condition);
+      eprosima::fastdds::dds::Condition * condition = &status_condition;
+      eprosima::fastdds::dds::GuardCondition * guard_condition =
+        &custom_event_info->get_listener()->event_guard[event->event_type];
+      bool active = false;
+      if (ReturnCode_t::RETCODE_OK == ret_code &&
+        triggered_coditions.end() != std::find_if(
+          triggered_coditions.begin(), triggered_coditions.end(),
+          [condition](const eprosima::fastdds::dds::Condition * c)
+          {
+            return c == condition;
+          }))
+      {
+        eprosima::fastdds::dds::Entity * entity = status_condition.get_entity();
+        eprosima::fastdds::dds::StatusMask changed_statuses = entity->get_status_changes();
+        if (changed_statuses.is_active(
+            rmw_fastrtps_shared_cpp::internal::rmw_event_to_dds_statusmask(
+              event->event_type)))
+        {
+          active = true;
+        }
```

Comment:
We could always do these checks, no need to check if the condition has been triggered.

Reply:
Done in 5ff6c5e2